### DISCUSSION TITLE
How to remove empty dir without suffix __XLDIR__

### DISCUSSION THREAD

**Discussion by loyalPlayer** [minio/minio#13113] [Q&A] [answered]
There are some empty dirs (without suffix __XLDIR__) on every data disks because of some unknow reasons. 
These dirs will be shown when execute "mc ls alias/".
But  they still left on device after execute "mc rm alias/{prefix}/empty_dir --recursive --force" although console had printed "Removing \`alias/{prefix}/empty_dir\`. " and finished.
I watched trace log and found DeleteMultipleObjects request and response 200 like normal directories.
I tried to cp objects to these dirs which cannot removed then I execute "mc rm alias/{prefix}/empty_dir --recursive --force" and it remove successfully.

## Expected Behavior

Remove successfully although it's a empty dir(without suffix __XLDIR__)

## Current Behavior

Cannot remove a empty dir(without suffix __XLDIR__)

## Steps to Reproduce (for bugs)

I don't know how to produce empty dir when cluster runs, but we can execute "mkdir -p"  to make a empty dir in the same bucket dir on all minio data disks to mock 

1.  mc mb alias/bucket1 
2.  execute command on every minio data disks 'mkdir -p {data_disk_path}/bucket1/emptydir' 
2.  mc ls alias/bucket1
It shows emptydir
3.  mc rm alias/bucket1/emptydir --recursive --force
Removing \`alias/bucket1/empty_dir\`. 
4.  mc ls alias/bucket1
emptydir is still existed.

## Context

1. We would do some business logic judgment by deal with information comes form listobjects.
2. These empty dir cannot clear by some legitimate approach
3. I want to know the reason why these empty dirs (without suffix __XLDIR__)  when cluster runs, I cannot find out it....
4. If it actually works as intended, please teach me why... A lot of thanks! I want to know more about it... and I will keep reading source code to find out...

## Your Environment

* Version used (`minio --version`):
2020-10-28 and 2021-03-17 have the same performance.
* Server setup and configuration:
* Operating System and version (`uname -a`):

**Comment by harshavardhana** [comment]
> There are some empty dirs (without suffix **XLDIR**) on every data disks because of some unknow reasons.
> These dirs will be shown when execute "mc ls alias/".
> But they still left on device after execute "mc rm alias/{prefix}/empty_dir --recursive --force" although console had printed "Removing `alias/{prefix}/empty_dir`. " and finished.
> I watched trace log and found DeleteMultipleObjects request and response 200 like normal directories.
> I tried to cp objects to these dirs which cannot removed then I execute "mc rm alias/{prefix}/empty_dir --recursive --force" and it remove successfully.

This should never happen in the first place @loyalPlayer - this just shows that you have a setup where your disks are either offline for too long or healing is not running as expected.

Please upgrade to our latest releases and then run `mc admin heal -r --remove` across the cluster to remove all these unexpected entries.

**Comment by loyalPlayer** [comment]
> > There are some empty dirs (without suffix **XLDIR**) on every data disks because of some unknow reasons.
> > These dirs will be shown when execute "mc ls alias/".
> > But they still left on device after execute "mc rm alias/{prefix}/empty_dir --recursive --force" although console had printed "Removing `alias/{prefix}/empty_dir`. " and finished.
> > I watched trace log and found DeleteMultipleObjects request and response 200 like normal directories.
> > I tried to cp objects to these dirs which cannot removed then I execute "mc rm alias/{prefix}/empty_dir --recursive --force" and it remove successfully.
> 
> This should never happen in the first place @loyalPlayer - this just shows that you have a setup where your disks are either offline for too long or healing is not running as expected.
> 
> Please upgrade to our latest releases and then run `mc admin heal -r --remove` across the cluster to remove all these unexpected entries.

Hi harshavardhana ~ Thank you for your reply.
Actually 'disk not found' comes up very frequently in server log.It would be helpful if you can tell me more detail about the setup you mentioned. I'm so confused about that (what is it, how it should be.)

**Comment by harshavardhana** [comment, marked as answer]
> Hi harshavardhana ~ Thank you for your reply.
> Actually 'disk not found' comes up very frequently in server log.It would be helpful if you can tell me more detail about the setup you mentioned. I'm so confused about that (what is it, how it should be.)

"disk not found" happens because you have perhaps a poor network setup when the servers are overloaded - you have to first work towards getting your network setup right. 

We provide such architecture guidance as part of our subscription https://min.io/pricing
